File: 22-generate-parentheses/22.py
class Solution:
    def generateParenthesis(self, n: int) -> List[str]:
        result = []

        def dfs(cur_str, left_count, right_count):
            if left_count == 0 and right_count == 0:
                result.append(cur_str)
                return
            
            if left_count > right_count:
                return

            if left_count > 0:
                dfs(cur_str + "(", left_count - 1, right_count)

            if right_count > 0:
                dfs(cur_str + ")", left_count, right_count - 1)

        dfs("", n, n)
        return result

File: 22-generate-parentheses/22.java
class Solution {
    public List<String> generateParenthesis(int n) {
        List<String> result = new ArrayList<String>();
        dfs("", n, n, result);
        return result;
    }

    public void dfs(String cur_str, int left_count, int right_count, List<String> result) {
        if (left_count == 0 && right_count == 0) {
            result.add(cur_str);
            return;
        }

        if (left_count > right_count) {
            return;
        }

        if (left_count > 0) {
            dfs(cur_str + "(", left_count - 1, right_count, result);
        }

        if (right_count > 0) {
            dfs(cur_str + ")", left_count, right_count - 1, result);
        }
    }
}
